conformance/lang/expressions/trap_expr.balt

```diff
@@ -0,0 +1,351 @@
+Labels: error-type, function-call-expr, function-defn, int, return-stmt, string, trap-expr, type-cast-expr, union-type
+
+function foo() returns int|string {
+    return "str";
+}
+
+function init() {
+    int|error t1 = trap <int>foo();
+    io:println(t1); // @output error("{ballerina}TypeCastError",message="incompatible types: 'string' cannot be cast to 'int'")
+    string|error t2 = trap <string>foo();
+    io:println(t2); // @output str
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a unary logical expression.
+Labels: boolean, error-type, function-call-expr, function-defn, int, return-stmt, trap-expr, union-type
+
+function foo() returns int|boolean {
+    return 23;
+}
+
+function init() {
+    boolean|error t1 = trap !<boolean>foo();
+    io:println(t1); // @output error("{ballerina}TypeCastError",message="incompatible types: 'int' cannot be cast to 'boolean'")
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a multiplicative expression.
+Labels: boolean, error-type, function-call-expr, function-defn, int, local-init-var-decl-stmt, multiplicative-expr, trap-expr, union-type, variable-reference-expr
+
+function init() {
+    int zero = 0;
+    int|error t1 = trap 2 / zero;
+    io:println(t1); // @output error("{ballerina}DivisionByZero",message=" / by zero")
+    int|error t2 = trap int:MAX_VALUE * int:MAX_VALUE;
+    io:println(t2); // @output error("{ballerina}NumberOverflow",message="int range overflow")
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a trap expression.
+Labels: boolean, error-type, function-call-expr, function-defn, int, local-init-var-decl-stmt, multiplicative-expr, trap-expr, union-type, variable-reference-expr
+
+function init() {
+    int zero = 0;
+    int|error t1 = trap trap 2 / zero;
+    io:println(t1); // @output error("{ballerina}DivisionByZero",message=" / by zero")
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a checking expression.
+Labels: boolean, check, checkpanic, error-type, int, local-init-var-decl-stmt, multiplicative-expr, trap-expr, union-type, variable-reference-expr
+
+function init() {
+    int zero = 0;
+    int|error e1 = trap (check 2 / zero);
+    io:println(e1); // @output error("{ballerina}DivisionByZero",message=" / by zero")
+    int|error e2 = trap (checkpanic 2 / zero);
+    io:println(e2); // @output error("{ballerina}DivisionByZero",message=" / by zero")
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a char and it does not panic.
+Labels: boolean, error-type, local-init-var-decl-stmt, string:Char, trap-expr, union-type, variable-reference-expr
+
+function init() {
+    string:Char c = "c";
+    string:Char|error e1 = trap c;
+    io:println(e1); // @output c
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a byte value and it does not panic.
```

Comment:
Can we add some tests with function calls here instead of literals and variable reference expressions that we know will never panic?

Reply:
@MaryamZi, do you mean here to add function calls that causes panic or add functions that return literal?

Reply:
Like a function call, which at compile-time we don't know if it will panic.

```ballerina
function init() {
    int|error a = trap fn("1");
    io:println(a);
}

function fn(string str) returns int {
    return checkpanic int:fromString(str);
}

```

Reply:
Changed